### ISSUE TITLE
Questions about your data

### ISSUE THREAD

**Issue by qdzwss** [Beckschen/TransUNet#23]
First of all, thank you for your contribution.I sent an email to you before and also downloaded your data set. I wanted to see your original data image, so I wrote a program to open your NPY data, but the image I got was all black, and the print also displayed 0 for the mdarray.What is the matter, please?

Also, if I want to train my own data set, is there anything I should pay attention to when creating it?

**Comment by Beckschen** [comment]
Hello, many thanks for your questions!  Sorry for the late reply.
1) 
1.1) Please note that the images saved in .npy are normalized. 
1.2) Some 2D slices would have no LABELS if those slices do not cover any organs. But I assume all IMAGE slices should have values.
1.3) Could you please kindly print like np.sum(image) to see if the data are all 0.

2) I think this repo is generalized well to other dataset. 
2.1) For 3D dataset, you could directly follow the code/setting of this repo. I assume most of the 3D CT scans have the shape (512, 512, Z).
If you want to run many dataset at the same time in one project, there is a interface for dataset configuration (See ['./train.py'](https://github.com/Beckschen/TransUNet/blob/main/train.py) line 58-64.)

2.2) The project also support 2D dataset since the network is a 2D network, and it can have outstanding performance as well if the scenario requires long-range attention like multi-organ.  If you want to train you own 2D dataset, I think you just have to modify the test script and pay attention to the augmentation since current project does not include random crop augmentation / random online scale.

Please let me know if you have specific questions if you start to implement your own dataset. I am willing to share my thoughts.

Thanks!

**Comment by qdzwss** [comment]
Hi,many thanks you for your answer.
1. I used np.sum() to find that the value of the data is not 0. This is my mistake
2. I have prepared my own 2D single channel bacterial data set with the size of 512*512. What I want to do is binary classification problem. Like your data, I converted the PNG file to NPZ file and modified the train.txt (in list_Synapse) and put them into the correct path.Here's where I changed the code:
  1.See './train.py' line 33--------&gt;default=512
  2.See './train.py' line 62--------&gt;'num_classes': 2,
  3.See './trainer.py' line 36--------&gt;num_workers=1

but i got the error:
(I think the reason is : nn.CrossEntropyLoss(input, target):The number of columns in input and target do not match
But I haven't found out how to change it so far. Could you help me?Thank you very much)

> ``load_pretrained: grid-size from 14 to 32
Namespace(base_lr=0.01, batch_size=5, dataset='Synapse', deterministic=1, exp='TU_Synapse512', img_size=512, is_pretrain=True, list_dir='./lists/lists_Synapse', max_epochs=50, max_iterations=30000, n_gpu=1, n_skip=3, num_classes=2, root_path='../data/Synapse/train_npz', seed=1234, vit_name='R50-ViT-B_16', vit_patches_size=16)
The length of train set is: 200
40 iterations per epoch. 2000 max iterations 
  0%|                                          | 0/50 [00:00<?, ?it/s]/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [9,0,0], thread: [256,0,0] Assertion `t >= 0 && t < n_classes` failed.
/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [416,0,0] Assertion `t >= 0 && t < n_classes` failed.
/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [417,0,0] Assertion `t >= 0 && t < n_classes` failed.
/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [288,0,0] Assertion `t >= 0 && t < n_classes` failed.
/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [289,0,0] Assertion `t >= 0 && t < n_classes` failed.
/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [8,0,0], thread: [64,0,0] Assertion `t >= 0 && t < n_classes` failed.

**Comment by minhoooo1** [comment]
> 首先，感谢您的贡献。我以前向您发送过一封电子邮件，还下载了您的数据集。我想看看你原来的数据图像，所以我写了一个程序来打开你的NPY数据，但我得到的图像是全黑的，打印也显示0为mdarray。请问怎么了？
> 
> 此外，如果我想训练自己的数据集，在创建数据集时有什么我应该注意的吗？

Hello!  Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you! 1121851272@qq.com

**Comment by fmd233** [comment]
Hello! Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you!1415451024@qq.com

**Comment by fmd233** [comment]
@qdzwss Hello! Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you!1415451024@qq.com

**Comment by q1031439207** [comment]
> Hi,many thanks you for your answer.
> 
> 1. I used np.sum() to find that the value of the data is not 0. This is my mistake
> 2. I have prepared my own 2D single channel bacterial data set with the size of 512*512. What I want to do is binary classification problem. Like your data, I converted the PNG file to NPZ file and modified the train.txt (in list_Synapse) and put them into the correct path.Here's where I changed the code:
>    1.See './train.py' line 33-------->default=512
>    2.See './train.py' line 62-------->'num_classes': 2,
>    3.See './trainer.py' line 36-------->num_workers=1
> 
> but i got the error:
> (I think the reason is : nn.CrossEntropyLoss(input, target):The number of columns in input and target do not match
> But I haven't found out how to change it so far. Could you help me?Thank you very much)
> 
> > ``load_pretrained: grid-size from 14 to 32
> > Namespace(base_lr=0.01, batch_size=5, dataset='Synapse', deterministic=1, exp='TU_Synapse512', img_size=512, is_pretrain=True, list_dir='./lists/lists_Synapse', max_epochs=50, max_iterations=30000, n_gpu=1, n_skip=3, num_classes=2, root_path='../data/Synapse/train_npz', seed=1234, vit_name='R50-ViT-B_16', vit_patches_size=16)
> > The length of train set is: 200
> > 40 iterations per epoch. 2000 max iterations
> > 0%|                                          | 0/50 [00:00<?, ?it/s]/pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [9,0,0], thread: [256,0,0] Assertion `t >= 0 && t < n_classes` failed.
> > /pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [416,0,0] Assertion `t >= 0 && t < n_classes` failed.
> > /pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [417,0,0] Assertion `t >= 0 && t < n_classes` failed.
> > /pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [288,0,0] Assertion `t >= 0 && t < n_classes` failed.
> > /pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [4,0,0], thread: [289,0,0] Assertion `t >= 0 && t < n_classes` failed.
> > /pytorch/aten/src/THCUNN/SpatialClassNLLCriterion.cu:104: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [8,0,0], thread: [64,0,0] Assertion `t >= 0 && t < n_classes` failed.

hello there, have you solved the problem？

**Comment by hyunwoo137** [comment]
Hello! Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you! hyunwoo137@naver.com

**Comment by q1031439207** [comment]
https://drive.google.com/drive/folders/1ACJEoTp-uqfFJ73qS3eUObQh52nGuzCd?usp=sharing



---Original---
From: ***@***.***&gt;
Date: Fri, Aug 13, 2021 16:22 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Beckschen/TransUNet] Questions about your data (#23)




 
Hello! Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you! ***@***.***
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by ssli23** [comment]
> First of all, thank you for your contribution.I sent an email to you before and also downloaded your data set. I wanted to see your original data image, so I wrote a program to open your NPY data, but the image I got was all black, and the print also displayed 0 for the mdarray.What is the matter, please?
> 
> Also, if I want to train my own data set, is there anything I should pay attention to when creating it?

Hello, I am sorry to bother you, can you send me the processed data, my email is shanshanli233@126.com

**Comment by AXu0511** [comment]
Could you be so kind to share the processed data? Looking forward to your reply.  My email is  wxl1009@mail.dlut.edu.cn.

**Comment by AXu0511** [comment]
> Hello! Cound you please send the preprocessed data set that can run the code to my mailbox? Thank you! [hyunwoo137@naver.com](mailto:hyunwoo137@naver.com)

Could you be so kind to share the processed data? Looking forward to your reply. My email is [wxl1009@mail.dlut.edu.cn]

**Comment by AXu0511** [comment]
> > First of all, thank you for your contribution.I sent an email to you before and also downloaded your data set. I wanted to see your original data image, so I wrote a program to open your NPY data, but the image I got was all black, and the print also displayed 0 for the mdarray.What is the matter, please?
> > Also, if I want to train my own data set, is there anything I should pay attention to when creating it?
> 
> Hello, I am sorry to bother you, can you send me the processed data, my email is [shanshanli233@126.com](mailto:shanshanli233@126.com)

Could you be so kind to share the processed data? Looking forward to your reply. My email is wxl1009@mail.dlut.edu.cn.

**Comment by AXu0511** [comment]
> First of all, thank you for your contribution.I sent an email to you before and also downloaded your data set. I wanted to see your original data image, so I wrote a program to open your NPY data, but the image I got was all black, and the print also displayed 0 for the mdarray.What is the matter, please?
> 
> Also, if I want to train my own data set, is there anything I should pay attention to when creating it?

Could you be so kind to share the processed data? Looking forward to your reply. My email is wxl1009@mail.dlut.edu.cn.

**Comment by zlm12345678** [comment]
> 你好，非常感谢你的提问！抱歉回复晚了。 1）1.1）请注意，保存在 .npy 中的图片是经过规范化处理的。 1.2）如果某些二维切片没有覆盖任何器官，则没有标签。但我认为所有IMAGE切片都应该有值。 1.3）请您打印 np.sum（image） 格式，看看数据是否全部为 0。
> 
> 2. 我觉得这个仓库在其他数据集上推广得很好。 2.1）对于3D数据集，你可以直接按照这个仓库的代码/设置操作。我猜大多数3D CT扫描都是这个形状（512、512、Z）。如果你想在一个项目中同时运行多个数据集，有一个数据集配置接口（见[“./train.py”](https://github.com/Beckschen/TransUNet/blob/main/train.py)第58-64行）。
> 
> 2.2）项目还支持二维数据集，因为网络是二维网络，如果场景需要远程关注（如多器官），也能表现出色。如果你想训练自己的二维数据集，我觉得你只需要修改测试脚本并关注增强，因为当前项目不包含随机裁剪增强/随机在线比例。
> 
> 如果你开始实施自己的数据集，如果有具体问题，请告诉我。我愿意分享我的想法。
> 
> 谢谢！

Hello, could you please provide the validation data and the `val.txt` file?
My email address is:qz956999@gmail.com
